File: src/models/MapModel.js
/**
 * MapModel.js - Handles map state and geographic data
 * Manages map layers, bounds, and coordinate transformations
 */
class MapModel {
  constructor() {
    this.mapCenter = [60.4518, 8.4689]; // Default: Norway center
    this.zoomLevel = 5;
    this.layers = new Map();
    this.activeFilters = {};
    this.spatialBounds = null;
  }

  /**
   * Add a layer to the map model
   * @param {string} name - Layer identifier
   * @param {object} config - Layer configuration
   */
  addLayer(name, config) {
    this.layers.set(name, {
      name,
      data: null,
      visible: true,
      ...config
    });
  }

  /**
   * Toggle layer visibility
   * @param {string} name - Layer identifier
   */
  toggleLayer(name) {
    if (this.layers.has(name)) {
      const layer = this.layers.get(name);
      layer.visible = !layer.visible;
      return layer.visible;
    }
    return false;
  }

  /**
   * Set spatial filter (e.g., bounding box)
   * @param {array} bounds - [[lat1, lng1], [lat2, lng2]]
   */
  setSpatialFilter(bounds) {
    this.spatialBounds = bounds;
  }

  /**
   * Get all visible layers
   * @returns {array} Visible layer objects
   */
  getVisibleLayers() {
    return Array.from(this.layers.values()).filter(l => l.visible);
  }

  /**
   * Update map viewport
   * @param {array} center - [latitude, longitude]
   * @param {number} zoom - Zoom level
   */
  setViewport(center, zoom) {
    this.mapCenter = center;
    this.zoomLevel = zoom;
  }

  /**
   * Store data for a layer
   * @param {string} layerName - Layer identifier
   * @param {object} data - GeoJSON or other geographic data
   */
  setLayerData(layerName, data) {
    if (this.layers.has(layerName)) {
      this.layers.get(layerName).data = data;
    }
  }
}

export default MapModel;

File: app/models/map_model.py
"""
MapModel.py - Manages map state, layers, and viewport
"""
from typing import Dict, List, Tuple, Optional


class MapModel:
    def __init__(self):
        self.map_center = [60.4518, 8.4689]  # Norway center
        self.zoom_level = 5
        self.layers = {}
        self.spatial_bounds = None
        self.search_point = None

    def add_layer(self, name: str, config: Dict):
        """
        Add a layer to the map model
        :param name: Layer identifier
        :param config: Layer configuration
        """
        self.layers[name] = {
            'name': name,
            'visible': True,
            'color': '#3388ff',
            'features': [],
            **config
        }

    def toggle_layer(self, name: str) -> bool:
        """Toggle layer visibility"""
        if name in self.layers:
            self.layers[name]['visible'] = not self.layers[name]['visible']
            return self.layers[name]['visible']
        return False

    def set_layer_features(self, name: str, features: List[Dict]):
        """Set features for a layer"""
        if name in self.layers:
            self.layers[name]['features'] = features

    def get_visible_layers(self) -> List[Dict]:
        """Get all visible layers"""
        return [layer for layer in self.layers.values() if layer['visible']]

    def set_viewport(self, center: List[float], zoom: int):
        """Update map viewport"""
        self.map_center = center
        self.zoom_level = zoom

    def set_search_point(self, lat: float, lng: float):
        """Set the center point for spatial search"""
        self.search_point = (lat, lng)

    def get_search_point(self) -> Optional[Tuple[float, float]]:
        """Get the search point"""
        return self.search_point
